contracts/transfer/src/tree.rs

```diff
@@ -4,189 +4,72 @@
 use phoenix_core::transaction::*;
-use phoenix_core::Note;
-use ranno::Annotation;
-
-#[derive(Debug, Clone)]
-struct TreeLeafWrapper(TreeLeaf);
-
-impl PoseidonLeaf for TreeLeafWrapper {
-    fn poseidon_hash(&self) -> BlsScalar {
-        rusk_abi::poseidon_hash(self.0.note.hash_inputs().into())
-    }
 
-    fn pos(&self) -> &u64 {
-        self.0.note.pos()
-    }
-
-    fn set_pos(&mut self, pos: u64) {
-        self.0.note.set_pos(pos);
-    }
-}
-
-impl Keyed<u64> for TreeLeafWrapper {
-    fn key(&self) -> &u64 {
-        &self.0.block_height
-    }
-}
-
-impl AsRef<Note> for TreeLeafWrapper {
-    fn as_ref(&self) -> &Note {
-        &self.0.note
-    }
-}
+use dusk_merkle::poseidon::{
+    Item as PoseidonItem, Opening as PoseidonOpening, Tree as PoseidonTree,
+};
 
-impl Borrow<u64> for TreeLeafWrapper {
-    fn borrow(&self) -> &u64 {
-        self.0.note.pos()
-    }
-}
+use crate::state::A;
 
-#[derive(Debug, Clone)]
 pub struct Tree {
-    tree: PoseidonTree<TreeLeafWrapper, u64, TRANSFER_TREE_DEPTH>,
+    tree: PoseidonTree<(), TRANSFER_TREE_DEPTH, A>,
+    // Since `dusk-merkle` does not include data blocks with the tree, we do it
+    // here.
+    leaves: Vec<TreeLeaf>,
 }
 
 impl Tree {
     pub const fn new() -> Self {
         Self {
             tree: PoseidonTree::new(),
+            leaves: Vec::new(),
         }
     }
 
     pub fn get(&self, pos: u64) -> Option<TreeLeaf> {
-        self.tree.get(pos).map(|l| l.0)
+        self.leaves.get(pos as usize).cloned()
     }
 
-    pub fn push(&mut self, leaf: TreeLeaf) -> u64 {
-        self.tree.push(TreeLeafWrapper(leaf))
+    pub fn push(&mut self, mut leaf: TreeLeaf) -> u64 {
+        // update the position before computing the hash
+        let pos = self.leaves.len() as u64;
+        leaf.note.set_pos(pos);
+
+        // compute the item that goes in the leaf of the tree
+        let hash = rusk_abi::poseidon_hash(leaf.note.hash_inputs().to_vec());
+        let item = PoseidonItem { hash, data: () };
+
+        self.tree.insert(pos, item);
+        self.leaves.push(leaf);
+
+        pos
     }
 
     pub fn root(&self) -> BlsScalar {
-        self.tree.root()
+        self.tree.root().hash
     }
 
-    pub fn leaves(
-        &self,
-        range: Range<u64>,
-    ) -> Option<impl Iterator<Item = &TreeLeaf>> {
-        self.tree
-            .annotated_iter_walk(HeightRangeWalker(range))
-            .map(|v| v.into_iter().map(|lw| &lw.0))
+    pub fn leaves(&self, range: Range<u64>) -> impl Iterator<Item = &TreeLeaf> {
+        // We can do this since we know the leaves are strictly increasing in
+        // block height. If this ever changes - such as in the case of a
+        // sparsely populated tree - we should annotate the tree and use
+        // `Tree::walk` instead.
+        self.leaves
+            .iter()
+            .skip_while(move |leaf| leaf.block_height < range.start)
+            .take_while(move |leaf| leaf.block_height < range.end)
```

Comment:
```suggestion
            .take_while(move |leaf| leaf.block_height <= range.end)
```

Reply:
A [range in Rust](https://doc.rust-lang.org/std/ops/struct.Range.html) is considered exclusive at the end. I would like to keep this convention.